Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,686 @@
+                code_lines.append(u"    %s.%s = %s() if %s is %s else %s" % (
+                    selfname, name, ph_name, name, default_factory_placeholder, name))
+            else:
+                # still need to use the default factory to initialize
+                code_lines.append(u"    %s.%s = %s()"
+                                  % (selfname, name, ph_name))
+    if node.scope.lookup("__post_init__"):
+        post_init_vars = ", ".join(name for name, field in fields.items()
+                                   if field.is_initvar)
+        code_lines.append("    %s.__post_init__(%s)" % (selfname, post_init_vars))
+    return u"\n".join(code_lines), placeholders, []
+
+
+def generate_repr_code(repr, node, fields):
+    """
+    The CPython implementation is just:
+    ['return self.__class__.__qualname__ + f"(' +
+                     ', '.join([f"{f.name}={{self.{f.name}!r}}"
+                                for f in fields]) +
+                     ')"'],
+
+    The only notable difference here is self.__class__.__qualname__ -> type(self).__name__
+    which is because Cython currently supports Python 2.
+    """
+    if not repr or node.scope.lookup("__repr__"):
+        return "", {}, []
+    code_lines = ["def __repr__(self):"]
+    strs = [ u"%s={self.%s}" % (name, name)
+            for name, field in fields.items() if field.repr.value and not field.is_initvar ]
+    format_string = u", ".join(strs)
+    code_lines.append(u"    return f'{type(self).__name__}(%s)'" % format_string)
+    code_lines = u"\n".join(code_lines)
+
+    return code_lines, {}, []
+
+def generate_cmp_code(op, funcname, node, fields):
+    if node.scope.lookup_here(funcname):
+        return "", {}, []
+
+    names = [ name for name, field in fields.items() if (field.compare.value and not field.is_initvar) ]
+
+    if not names:
+        return "", {}, []  # no comparable types
+
+    code_lines = [
+        "def %s(self, other):" % funcname,
+        "    cdef %s other_cast" % node.class_name,
+        "    if isinstance(other, %s):" % node.class_name,
+        "        other_cast = <%s>other" % node.class_name,
+        "    else:",
+        "        return NotImplemented"
+    ]
+
+    # The Python implementation of dataclasses.py does a tuple comparison
+    # (roughly):
+    #  return self._attributes_to_tuple() {op} other._attributes_to_tuple()
+    #
+    # For the Cython implementation a tuple comparison isn't an option because
+    # not all attributes can be converted to Python objects and stored in a tuple
+    #
+    # TODO - better diagnostics of whether the types support comparison before
+    #    generating the code. Plus, do we want to convert C structs to dicts and
+    #    compare them that way (I think not, but it might be in demand)?
+    checks = []
+    for name in names:
+        checks.append("(self.%s %s other_cast.%s)" % (
+            name, op, name))
+    if checks:
```

Comment:
When you do it like this they look connected

```suggestion
            name, op, name))

    if checks:
```